revlist/revlist.go

```diff
@@ -0,0 +1,272 @@
+// contents than what can be found in "c".
+func differentContents(path string, c *git.Commit, cs []*git.Commit) ([]*git.Commit, error) {
+	result := make([]*git.Commit, 0, len(cs))
+	h, found := blobHash(path, c)
+	if !found {
+		return nil, git.ErrFileNotFound
+	}
+	for _, cx := range cs {
+		if hx, found := blobHash(path, cx); found && h != hx {
+			result = append(result, cx)
+		}
+	}
+	return result, nil
+}
+
+// blobHash returns the hash of a path in a commit
+func blobHash(path string, commit *git.Commit) (hash core.Hash, found bool) {
+	file, err := commit.File(path)
+	if err != nil {
+		var empty core.Hash
+		return empty, found
+	}
+	return file.Hash, true
+}
+
+// Returns a new slice of commits, with duplicates removed.  Expects a
+// sorted commit list.  Duplication is defined according to "comp".  It
+// will always keep the first commit of a series of duplicated commits.
+func removeComp(path string, cs []*git.Commit, comp func(string, *git.Commit, *git.Commit) (bool, error)) ([]*git.Commit, error) {
+	result := make([]*git.Commit, 0, len(cs))
+	if len(cs) == 0 {
+		return result, nil
+	}
+	result = append(result, cs[0])
+	for i := 1; i < len(cs); i++ {
+		equals, err := comp(path, cs[i], cs[i-1])
+		if err != nil {
+			return nil, err
+		}
+		if !equals {
+			result = append(result, cs[i])
+		}
+	}
+	return result, nil
+}
+
+// Equivalent commits are commits whos patch is the same.
+func equivalent(path string, a, b *git.Commit) (bool, error) {
+	numParentsA := a.NumParents()
+	numParentsB := b.NumParents()
+
+	// the first commit is not equivalent to anyone
+	// and "I think" merges can not be equivalent to anything
+	if numParentsA != 1 || numParentsB != 1 {
+		return false, nil
+	}
+
+	diffsA, err := patch(a, path)
+	if err != nil {
+		return false, err
+	}
+	diffsB, err := patch(b, path)
+	if err != nil {
+		return false, err
+	}
+
+	return sameDiffs(diffsA, diffsB), nil
+}
+
+func patch(c *git.Commit, path string) ([]diffmatchpatch.Diff, error) {
+	// get contents of the file in the commit
+	file, err := c.File(path)
+	if err != nil {
+		return nil, err
+	}
+	content := file.Contents()
+
+	// get contents of the file in the first parent of the commit
+	var contentParent string
+	iter := c.Parents()
+	parent, err := iter.Next()
+	if err != nil {
+		return nil, err
+	}
+	file, err = parent.File(path)
+	if err != nil {
+		contentParent = ""
+	} else {
+		contentParent = file.Contents()
+	}
+
+	// compare the contents of parent and child
+	return diff.Do(content, contentParent), nil
+}
+
+func sameDiffs(a, b []diffmatchpatch.Diff) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if !sameDiff(a[i], b[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func sameDiff(a, b diffmatchpatch.Diff) bool {
+	if a.Type != b.Type {
+		return false
+	}
+	switch a.Type {
+	case 0:
+		return git.CountLines(a.Text) == git.CountLines(b.Text)
```

Comment:
Yes, we don't care about the contents of the lines only if the number of "untouched" ones are the same, that is what make to diffs equivalents.